### PR TITLE
Extend add command with time support

### LINKED ISSUE #30: Deadline support for add command
So that I can keep track of events with defined start and end dates (or deadlines)
Similarly, I should be able to edit and remove start/end time/deadlines so that I can respond to changes accordingly.
- Implement DateTimeParser
- Implement Schedule Command

Note
- Cater for the possibility of task description containing keywords such as ‘from’ ‘to’. e.g., The user can  add the task ‘watch day after tomorrow’ without the software interpreting ‘tomorrow’ as the day for the task.
- Support floating tasks, tasks with deadlines and event with start and end times
- Consider using PrettyTime time parser: http://www.ocpsoft.org/prettytime/nlp/

### PR DESCRIPTION
**Pull request by burnflare**
Resolves #30 
TODO:
- [x] Only works with BOTH `/from` and `/to` commands. Need to tweak the regex to fix that
- [x] Support `/by` after fixing above bug
- [x] Add test cases - floating tasks, task with a deadline, events with a start and end time, invalid time formats
- [x] Update XmlAdaptedTask to save the task start and end date time
- [ ] Update add command message usage/error
- [x] Comment the acceptable time format on the PR for others to test
- [x] If today is Monday and the command is write "Add something from Wed 10 am to 12pm" it should parse as: From: Wednesday 10am To: Wednesday 12pm. Instead of To: Monday 12pm
- [x] whenever a day is not specified, it is taken as today. Another bad example, "add something from 10pm to 12pm", end time will be before start time
- [x] Allow user to add command "drop by school" without throwing an invalid add/date time format exception. (Options: use -by or /by, ignore words enclosed within inverted commas, check if word after by is a time)
- [x] implement 'schedule' command (refer to user guide)

### CODE REVIEW

File: src/main/java/seedu/agendum/logic/parser/Parser.java [hunk 1]
```diff
@@ -95,16 +98,35 @@ public Command parseCommand(String userInput) {
...
+
         try {
-            return new AddCommand(
-                    matcher.group("name"),
-                    getTagsFromArgs(matcher.group("tagArguments"))
-            );
+            String name = matcher.group("name");
+            Optional<LocalDateTime> byDateTime = DateTimeParser.parseString("");
+            Optional<LocalDateTime> fromDateTime = DateTimeParser.parseString(matcher.group("fromArg"));
+            Optional<LocalDateTime> toDateTime = DateTimeParser.parseString(matcher.group("toArg"));
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/parser/Parser.java]: Better guarding here

File: src/main/java/seedu/agendum/logic/parser/Parser.java [hunk 2]
```diff
@@ -95,16 +98,35 @@ public Command parseCommand(String userInput) {
...
+
+            if (byDateTime.isPresent()) {
+                return new AddCommand(
+                        name,
+                        byDateTime,
+                        getTagsFromArgs("test")
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/parser/Parser.java]: Temporary to placate the constructor. An upcoming PR that removes tagging support will clean this up

File: src/main/java/seedu/agendum/logic/parser/Parser.java [hunk 3]
```diff
@@ -32,6 +33,8 @@
 
     private static final Pattern RENAME_ARGS_FORMAT = Pattern.compile("(?<targetIndex>\\d+)\\s+(?<name>[^/]+)");
 
+    private static final Pattern ADD_ARGS_FORMAT = Pattern.compile("(?<name>[^/]+)(\\/from\\s)(?<fromArg>\\b.+?(?=\\/|$)){0,1}(\\/to\\s)(?<toArg>\\b.+?(?=\\/|$)){0,1}");
```
Comment (burnflare) [inline src/main/java/seedu/agendum/logic/parser/Parser.java]: Regex not done, has bugs. Check PR description

Comment (burnflare) [reply src/main/java/seedu/agendum/logic/parser/Parser.java]: Problem with the regex is that it's not taking the 2 regex's as optionals, even thought I ended it with a `{0,1}`. So if  it only finds a `/from` and not a `/to`, matcher.matches() fails.

That should not be the case. We need them to be optional. i.e. you could enter `/from`, `/to` or nothing at all.

**Comment by burnflare** [comment]
This is ready for review. I've made tasks for the outstanding features as issues.

For testing, try any kind of format that you can think of.

**Comment by burnflare** [comment]
Someone approve this PR?

File: src/test/java/seedu/agendum/logic/DateTimerParserTest.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    private void assertDates(LocalDateTime dateTime1, LocalDateTime dateTime2) {
+        assertEquals(dateTime1, dateTime2);
+    }
+
+    @Test
+    public void simpleDateTest() throws Exception {
+        Optional<LocalDateTime> t = DateTimeParser.parseString("2016/01/01 00:00");
+        assertDates(t.get(), LocalDateTime.of(2016,1,1,0,0));
+    }
+}
```
Comment (INCENDE) [inline src/test/java/seedu/agendum/logic/DateTimerParserTest.java]: Might need to add more tests

**Review by INCENDE — APPROVED** [review]
Looks good

**Comment by burnflare** [comment]
Good point, will work on it.

**Pull request merged**